tensorflow/core/kernels/fused_embedding/embedding_lookup_sparse_pre_op.cc

```diff
@@ -0,0 +1,314 @@
+    // 1. get input tensor
+    Tensor const* values_tensor = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->input("sp_values", &values_tensor));
+    const int64_t nnz = values_tensor->shape().dim_size(0);
+
+    const int64_t* values = reinterpret_cast<const int64_t*>(
+                                  values_tensor->flat<int64>().data());
+
+    Tensor const* indices_tensor = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->input("sp_indices", &indices_tensor));
+
+    const int64_t* indices = reinterpret_cast<const int64_t*>(
+                                  indices_tensor->flat<int64>().data());
+
+    Tensor const* dense_shape = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->input("sp_dense_shape", &dense_shape));
+    const int64_t batch_size = dense_shape->flat<int64>().data()[0];
+
+    OpInputList partition_shapes;
+    OP_REQUIRES_OK(ctx, ctx->input_list("partition_shapes", &partition_shapes));
+
+    partition_total_sizes_ = 0;
+    for (const Tensor& shape : partition_shapes) {
+      OP_REQUIRES(ctx, shape.dims() <= 2,
+                  errors::InvalidArgument(
+                      "input partition_shapes must all less than rank 2"));
+      partition_total_sizes_ += shape.flat<int64>().data()[0];
+    }
+
+    // fixme(marvin): show error info when got fake input.
+    OP_REQUIRES(ctx, partition_total_sizes_ != 1,
+        errors::InvalidArgument("Not support EV yet"));
+
+    // 1.1 define output tensors
+    OpOutputList partitioned_values;
+    OP_REQUIRES_OK(ctx,
+                   ctx->output_list("partitioned_values", &partitioned_values));
+    OpOutputList partitioned_indices;
+    OP_REQUIRES_OK(
+        ctx, ctx->output_list("partitioned_indices", &partitioned_indices));
+
+    Tensor* all_flags;
+    OP_REQUIRES_OK(
+        ctx, ctx->allocate_output(2 * num_partitions_,
+                                  TensorShape{batch_size + nnz}, &all_flags));
+    int32_t* all_flags_list = all_flags->flat<int32_t>().data();
+
+    memset(all_flags_list, 0, (batch_size + nnz) * sizeof(int32_t));
+    ShowLog(start, "// 1.1 define output tensors");
+
+    // 2.1 get index
+    const int64_t idsPerPartition = partition_total_sizes_ / num_partitions_;
+    const int64_t extras = partition_total_sizes_ % num_partitions_;
+    std::vector<int64_t> empty_index_;
+    // [p_seg_nums + list(p_seg, p_id)]
+    int64_t* const id_index_array = new int64_t[num_partitions_ + 1 + nnz * 2];
+    memset(id_index_array, 0, (num_partitions_ + 1) * sizeof(int64_t));
+    ShowLog(start, "// 1.2 memset output");
+
+    // 2.2 get the map of the mutli-table index
+    int64_t default_p_seg = 0;
+    int64_t default_p_val = 0;
+    int64_t p_seg = 0;
+    int64_t p_val = 0;
+    register int64_t tmp_id;
+    int64_t* const min_id_per_seg = new int64_t[num_partitions_];
+#ifdef __AVX512F__
+    int64_t* tmp_value_arr;
+
+    // 2.1 build min_id_per_seg
+    memset(min_id_per_seg, 0, (num_partitions_) * sizeof(int64_t));
+    for(int i = 0; i < num_partitions_; ++i){
```

Comment:
format, blank after "for", pls check all places.

Reply:
Done.